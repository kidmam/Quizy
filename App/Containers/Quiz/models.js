export function Question (payload) {
  return {
    category: payload.category,
    type: payload.type,
    difficulty: payload.difficulty,
    question: payload.question,
    correctAnswer: payload.correct_answer === 'True',
    isCorrect: null
  }
}
