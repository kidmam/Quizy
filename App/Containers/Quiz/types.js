// @flow

export type Question = {
  category: string,
  type: string,
  difficulty: string,
  question: string,
  correctAnswer: boolean,
  isCorrect: ?boolean
}

export type QuizState = {
  results: Array<Question>,
  isFetching: boolean,
  error: boolean,
  currentQuestionIndex: number
}

export function parseQuestion (payload: Object): Question {
  return {
    category: payload.category,
    type: payload.type,
    difficulty: payload.difficulty,
    question: payload.question,
    correctAnswer: payload.correct_answer === 'True',
    isCorrect: null
  }
}
