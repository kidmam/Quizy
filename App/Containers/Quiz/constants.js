export const FETCH_QUESTIONS = 'Quizy/Quiz/FETCH_QUESTIONS'
export const FETCH_QUESTIONS_PENDING = 'Quizy/Quiz/FETCH_QUESTIONS_PENDING'
export const FETCH_QUESTIONS_FULFILLED = 'Quizy/Quiz/FETCH_QUESTIONS_FULFILLED'
export const FETCH_QUESTIONS_REJECTED = 'Quizy/Quiz/FETCH_QUESTIONS_REJECTED'

export const ANSWER_QUESTION = 'Quizy/Quiz/ANSWER_QUESTION'

export const CLEAR_QUESTIONS = 'Quizy/Quiz/CLEAR_QUESTIONS'
