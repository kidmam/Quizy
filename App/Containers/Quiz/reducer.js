import Immutable from 'seamless-immutable'
import {
  FETCH_QUESTIONS_PENDING,
  FETCH_QUESTIONS_FULFILLED,
  FETCH_QUESTIONS_REJECTED,
  ANSWER_QUESTION,
  CLEAR_QUESTIONS
} from './constants'
import { parseQuestion, QuizState } from './models'

const INITIAL_STATE: QuizState = Immutable({
  results: [],
  isFetching: false,
  error: false,
  currentQuestionIndex: 0
})

export default function reducer (state = INITIAL_STATE, action) {
  switch (action.type) {
    case FETCH_QUESTIONS_PENDING:
      return {
        ...state,
        results: [],
        isFetching: true
      }
    case FETCH_QUESTIONS_FULFILLED:
      return {
        ...state,
        isFetching: false,
        results: action.payload.data.results.map((result) => parseQuestion(result))
      }
    case FETCH_QUESTIONS_REJECTED:
      return {
        ...state,
        isFetching: false,
        error: true
      }
    case ANSWER_QUESTION:
      const results = state.results.map((result, index) => {
        if (index === state.currentQuestionIndex) {
          result.isCorrect = action.payload === result.correctAnswer
        }
        return result
      })
      const isLastQuestion = state.results.length === state.currentQuestionIndex + 1
      return {
        ...state,
        results,
        currentQuestionIndex: !isLastQuestion ? state.currentQuestionIndex + 1 : state.currentQuestionIndex
      }
    case CLEAR_QUESTIONS:
      return INITIAL_STATE
    default:
      return state
  }
}
