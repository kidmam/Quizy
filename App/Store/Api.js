import OpenTDBApi from '../Services/OpenTDBApi'

export const openTDBApi = OpenTDBApi.create()
