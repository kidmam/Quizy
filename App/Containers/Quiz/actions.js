import { FETCH_QUESTIONS, ANSWER_QUESTION, CLEAR_QUESTIONS } from './constants'
import { openTDBApi } from '../../Store/Api'

export function fetchQuestions () {
  return {
    type: FETCH_QUESTIONS,
    payload: openTDBApi.getQuestions()
  }
}

export function answerQuestion (answer) {
  return {
    type: ANSWER_QUESTION,
    payload: answer
  }
}

export function clearQuestions () {
  return {
    type: CLEAR_QUESTIONS
  }
}
