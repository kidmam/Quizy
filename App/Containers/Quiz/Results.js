// @flow

import React from 'react'
import { View, Text, StyleSheet, FlatList } from 'react-native'
import { connect } from 'react-redux'
import { NavigationActions } from 'react-navigation'
import entities from 'entities'
import { fetchQuestions, clearQuestions } from './actions'
import Button from '../../Components/Button'
import { Colors, Fonts } from '../../Themes'
import type { QuizState } from './models'

type ResultsComponent = {
  clearQuestions: Function,
  fetchQuestions: Function,
  quiz: QuizState,
  navigation: Object
}

class Results extends React.PureComponent<ResultsComponent> {
  reset = () => {
    this.props.clearQuestions()
    this.props.fetchQuestions()
    const resetAction = NavigationActions.reset({
      index: 0,
      actions: [NavigationActions.navigate({ routeName: 'Intro' })]
    })
    this.props.navigation.dispatch(resetAction)
  }

  getScore = () => this.props.quiz.results.filter((result) => result.isCorrect)

  renderItem ({ item }) {
    return (
      <View style={[styles.questionContainer, item.isCorrect ? styles.questionContainerCorrect : null]}>
        <Text style={styles.question}>{entities.decodeHTML(item.question)}</Text>
        <Text style={styles.answer}>{item.correctAnswer ? 'TRUE' : 'FALSE'}</Text>
      </View>
    )
  }

  render () {
    const score = this.getScore()
    return (
      <View style={styles.container}>
        <View style={styles.welcomeContainer}>
          <Text style={styles.welcome}>Your scored</Text>
          <Text style={styles.welcome}>
            {score.length} / {this.props.quiz.results.length}
          </Text>
          <FlatList
            keyExtractor={(_, index) => `result__${index}`}
            data={this.props.quiz.results}
            renderItem={this.renderItem}
          />
        </View>
        <Button text='PLAY AGAIN' onPress={this.reset} />
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.background
  },
  welcomeContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    paddingTop: 20,
    paddingBottom: 10
  },
  welcome: {
    ...Fonts.style.h1,
    color: Colors.textAccent,
    textAlign: 'center'
  },
  questionContainer: { backgroundColor: Colors.red, paddingVertical: 10, marginTop: 5, paddingHorizontal: 30 },
  question: {
    ...Fonts.style.normal,
    color: Colors.textAccent
  },
  answer: {
    ...Fonts.style.normal,
    marginTop: 10,
    fontWeight: 'bold',
    color: Colors.textAccent,
    textAlign: 'center'
  },
  questionContainerCorrect: {
    backgroundColor: Colors.green
  }
})

const mapStateToProps = (state) => ({ quiz: state.quiz })

const mapDispatchToProps = (dispatch) => ({
  fetchQuestions: () => dispatch(fetchQuestions()),
  clearQuestions: () => dispatch(clearQuestions())
})

export default connect(mapStateToProps, mapDispatchToProps)(Results)
