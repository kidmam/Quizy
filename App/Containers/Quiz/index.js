// @flow

import React, { Component } from 'react'
import { View, Text, StyleSheet, TouchableOpacity, ActivityIndicator } from 'react-native'
import { connect } from 'react-redux'
import Icon from 'react-native-vector-icons/FontAwesome'
import PropTypes from 'prop-types'
import entities from 'entities'
import { fetchQuestions, answerQuestion } from './actions'
import { Colors, Fonts } from '../../Themes'

class Quiz extends Component {
  componentDidMount () {
    this.props.fetchQuestions()
  }

  finish = () => {
    this.props.navigation.navigate('QuizResults')
  }

  getCurrentResult = () => this.props.quiz.results[this.props.quiz.currentQuestionIndex]

  answerQuestion = (answer) => {
    this.props.answerQuestion(answer)
    if (this.props.quiz.results.length === this.props.quiz.currentQuestionIndex + 1) {
      this.finish()
    }
  }

  renderContent () {
    if (this.props.quiz.isFetching || this.props.quiz.results.length === 0) {
      return (
        <View style={styles.activityIndicatorContainer}>
          <ActivityIndicator color={Colors.primary} />
        </View>
      )
    }

    const result = this.getCurrentResult()

    return (
      <View style={styles.content}>
        <View style={styles.categoryContainer}>
          <Text style={styles.category}>{result.category}</Text>
        </View>
        <View style={styles.questionContainer}>
          <Text style={styles.question}>{entities.decodeHTML(result.question)}</Text>
        </View>
        <Text style={styles.currentIndex}>
          {this.props.quiz.currentQuestionIndex + 1} of {this.props.quiz.results.length}
        </Text>
        <View style={styles.actionsSection}>
          <TouchableOpacity
            style={[styles.button, styles.buttonFalse]}
            onPress={() => {
              this.answerQuestion(false)
            }}
          >
            <Icon name='times' size={40} color='white' />
          </TouchableOpacity>
          <TouchableOpacity
            style={[styles.button, styles.buttonTrue]}
            onPress={() => {
              this.answerQuestion(true)
            }}
          >
            <Icon name='check' size={40} color='white' />
          </TouchableOpacity>
        </View>
      </View>
    )
  }

  render () {
    return <View style={styles.container}>{this.renderContent()}</View>
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.background
  },
  activityIndicatorContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center'
  },
  content: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center'
  },
  categoryContainer: {
    justifyContent: 'center',
    alignItems: 'center',
    marginBottom: 40,
    paddingHorizontal: 20,
    minHeight: 120
  },
  category: {
    ...Fonts.style.h2,
    color: Colors.textAccent,
    textAlign: 'center'
  },
  questionContainer: {
    width: 250,
    height: 200,
    alignItems: 'center',
    justifyContent: 'center',
    borderColor: Colors.primary,
    borderWidth: 1,
    padding: 15,
    borderRadius: 5
  },
  question: {
    ...Fonts.style.normal,
    maxWidth: 250,
    textAlign: 'center',
    color: Colors.textAccent
  },
  actionsSection: {
    flexDirection: 'row',
    marginTop: 20,
    justifyContent: 'space-between'
  },
  button: {
    width: 80,
    height: 80,
    borderRadius: 40,
    justifyContent: 'center',
    alignItems: 'center',
    marginHorizontal: 45
  },
  buttonFalse: {
    backgroundColor: Colors.black
  },
  buttonTrue: {
    backgroundColor: Colors.button
  },
  currentIndex: {
    ...Fonts.style.normal,
    textAlign: 'center',
    color: Colors.textAccent,
    marginTop: 20
  }
})

Quiz.propTypes = {
  fetchQuestions: PropTypes.func.isRequired,
  answerQuestion: PropTypes.func.isRequired,
  quiz: PropTypes.object.isRequired
}

const mapStateToProps = (state) => ({ quiz: state.quiz })

const mapDispatchToProps = (dispatch) => ({
  fetchQuestions: () => dispatch(fetchQuestions()),
  answerQuestion: (answer) => dispatch(answerQuestion(answer))
})

export default connect(mapStateToProps, mapDispatchToProps)(Quiz)
