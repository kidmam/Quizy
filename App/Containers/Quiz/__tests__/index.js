import 'react-native'
import React from 'react'
import thunk from 'redux-thunk'
import sinon from 'sinon'
import renderer from 'react-test-renderer'
import Quiz from '../index'
import createStore from '../../../Store/Redux'
import { openTDBApi } from '../../../Store/Api'
import * as types from '../constants'
import { parseQuestion } from '../types'

const questions = {
  data: {
    results: [
      parseQuestion({
        category: 'Entertainment: Video Games',
        type: 'multiple',
        difficulty: 'easy',
        question: 'What year was the game Team Fortress 2 released?',
        correct_answer: '2007'
      }),
      parseQuestion({
        category: 'Geography',
        type: 'multiple',
        difficulty: 'medium',
        question: 'In which city, is the Big Nickel located in Canada?',
        correct_answer: 'Sudbury, Ontario'
      })
    ]
  }
}

describe('Quiz Container', () => {
  let store

  beforeEach(() => {
    sinon.stub(openTDBApi, 'getQuestions')
    openTDBApi.getQuestions.returns(new Promise((resolve) => resolve(questions)))
    store = createStore(false)
  })

  afterEach(() => {
    openTDBApi.getQuestions.restore()
  })

  it('Renders Quiz', async () => {
    const QuizComponent = renderer.create(<Quiz store={store} />)
    expect(QuizComponent).toMatchSnapshot()
  })

  it('Starts with question #1', () => {
    renderer.create(<Quiz store={store} />)
    const state = store.getState()
    expect(state.quiz.currentQuestionIndex + 1).toEqual(1)
  })

  it('Fetchs the questions', () => {
    const QuizComponent = renderer.create(<Quiz store={store} />)
    store.dispatch({
      type: types.FETCH_QUESTIONS_FULFILLED,
      payload: questions
    })
    const state = store.getState()
    expect(state.quiz.results).toEqual(questions.data.results)
  })

  it('Increases the currentQuestionIndex onPress the YES button', () => {
    const QuizComponent = renderer.create(<Quiz store={store} />)
    store.dispatch({
      type: types.FETCH_QUESTIONS_FULFILLED,
      payload: questions
    })
    QuizComponent.root.findByProps({ id: 'answerButtonYes' }).props.onPress()
    expect(store.getState().quiz.currentQuestionIndex).toEqual(1)
  })

  it('Increases the currentQuestionIndex onPress the NO button', () => {
    const QuizComponent = renderer.create(<Quiz store={store} />)
    store.dispatch({
      type: types.FETCH_QUESTIONS_FULFILLED,
      payload: questions
    })
    QuizComponent.root.findByProps({ id: 'answerButtonYes' }).props.onPress()
    expect(store.getState().quiz.currentQuestionIndex).toEqual(1)
  })

  it('Restart the state when the component renders', () => {
    const QuizComponent = renderer.create(<Quiz store={store} />)
    store.dispatch({
      type: types.FETCH_QUESTIONS_FULFILLED,
      payload: questions
    })
    QuizComponent.root.findByProps({ id: 'answerButtonYes' }).props.onPress()
    expect(store.getState().quiz.currentQuestionIndex).toEqual(1)
    const QuizComponent2 = renderer.create(<Quiz store={store} />)
    expect(store.getState().quiz.currentQuestionIndex).toEqual(0)
  })
})
