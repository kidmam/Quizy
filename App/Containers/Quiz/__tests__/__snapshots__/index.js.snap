// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz Container Renders Quiz 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#7D59C8",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <ActivityIndicator
      animating={true}
      color="#FFFFFF"
      hidesWhenStopped={true}
      size="small"
    />
  </View>
</View>
`;
